File: php/leet_code/623.php
<?php

/**
 * This file contains the TreeNode and Solution classes.
 *
 * PHP version 8.3
 *
 * @category LeetCode
 * @package  LeetCode\Solution
 * @author   Jerodg
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     http://github.com/jerodg/leet_code
 */

declare(strict_types=1);

/**
 * The TreeNode class represents a node in a binary tree.
 *
 * @category LeetCode
 * @package  LeetCode\Solution
 * @author   Jerodg
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     http://github.com/jerodg/leet_code
 */
class TreeNode
{
    public $val = null;
    public $left = null;
    public $right = null;

    /**
     * TreeNode constructor.
     *
     * @param int|null $val The value of the node.
     * @param TreeNode|null $left The left child of the node.
     * @param TreeNode|null $right The right child of the node.
     */
    public function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

/**
 * The Solution class contains methods for solving specific problems related to binary trees.
 *
 * @category LeetCode
 * @package  LeetCode\Solution
 * @author   Jerodg
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     http://github.com/jerodg/leet_code
 */
class Solution
{
    /**
     * Adds a row of nodes with a given value at a given depth in a binary tree.
     *
     * This method takes a binary tree, a value, and a depth, and adds a row of nodes with the given value at the given depth in the tree.
     *
     * @param TreeNode|null $root The root of the binary tree.
     * @param int $val The value of the nodes in the new row.
     * @param int $depth The depth at which to add the new row.
     *
     * @return TreeNode|null The root of the binary tree after adding the new row.
     *
     * @throws InvalidArgumentException If $depth is less than 1.
     *
     * @example
     * $solution = new Solution();
     * $root = new TreeNode(4, new TreeNode(2, new TreeNode(3), new TreeNode(1)), new TreeNode(6, new TreeNode(5)));
     * $newRoot = $solution->addOneRow($root, 1, 2);
     * // The binary tree now looks like this:
     * //     4
     * //   /   \
     * //  1     1
     * // /       \
     * //2         6
     * // / \     /
     * //3   1   5
     */
    public function addOneRow(?TreeNode $root, int $val, int $depth): ?TreeNode
    {
        if ($depth === 1) {
            return new TreeNode($val, $root, null);
        }
        if (!$root) {
            return [];
        }

        $res = $root;
        $sample = new TreeNode($val);
        $arr = [$root];
        $counter = 2;
        while (count($arr) > 0) {
            $tmp = [];
            foreach ($arr as $v) {
                if ($counter === $depth) {
                    $left = clone $sample;
                    $right = clone $sample;
                    if ($v->left) {
                        $left->left = $v->left;
                    }
                    if ($v->right) {
                        $right->right = $v->right;
                    }
                    $v->left = $left;
                    $v->right = $right;
                }
                if ($v->left) {
                    $tmp[] = $v->left;
                }
                if ($v->right) {
                    $tmp[] = $v->right;
                }
            }
            $arr = $tmp;
            if ($counter === $depth) {
                break;
            }

            $counter++;
        }
        return $res;
    }
}

File: ruby/leet_code/623.rb
# Class TreeNode
#
# This class represents a node in a binary tree. Each node has a value, and references to its left and right children.
#
class TreeNode
  attr_accessor :val, :left, :right

  # Initialize a new instance of the TreeNode class
  #
  # @param val [Integer] The value of the node
  # @param left [TreeNode, nil] The left child of the node
  # @param right [TreeNode, nil] The right child of the node
  #
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

# Function to add a row of nodes with a given value at a given depth in the tree
#
# @param root [TreeNode] The root of the tree
# @param val [Integer] The value for the new row of nodes
# @param depth [Integer] The depth at which to add the new row
# @return [TreeNode] The root of the modified tree
#
# @example
#   add_one_row(root, 1, 2) #=> root with an added row of nodes with value 1 at depth 2
#
def add_one_row(root, val, depth)
  if depth == 1
    new_root = TreeNode.new(val)
    new_root.left = root
    return new_root
  end

  dfs_insert(root, val, depth, 1)
  root
end

# Function to insert a new row of nodes at a given depth in the tree using depth-first search
#
# @param node [TreeNode] The current node
# @param val [Integer] The value for the new row of nodes
# @param target_depth [Integer] The depth at which to add the new row
# @param current_depth [Integer] The current depth in the tree
#
# @example
#   dfs_insert(root, 1, 2, 1) #=> root with an added row of nodes with value 1 at depth 2
#
def dfs_insert(node, val, target_depth, current_depth)
  return unless node
  return if current_depth == target_depth

  if current_depth == target_depth - 1
    old_left = node.left
    old_right = node.right

    node.left = TreeNode.new(val)
    node.right = TreeNode.new(val)

    node.left.left = old_left
    node.right.right = old_right
  else
    dfs_insert(node.left, val, target_depth, current_depth + 1)
    dfs_insert(node.right, val, target_depth, current_depth + 1)
  end
end